lib/backend/k8s/syncer.go

```diff
@@ -1047,20 +1015,55 @@ func (syn *kubeSyncer) parseGlobalConfigEvent(e watch.Event) *model.KVPair {
 	return kvp
 }
 
+func (syn *kubeSyncer) parseSystemNetworkPolicyEvent(e watch.Event) *model.KVPair {
+	return syn.parseCustomK8sResourceEvent(e, resources.SystemNetworkPolicyConverter{}, "SystemNetworkPolicy")
+}
+
 func (syn *kubeSyncer) parseIPPoolEvent(e watch.Event) *model.KVPair {
-	log.Debug("Parsing IPPool watch event")
+	return syn.parseCustomK8sResourceEvent(e, resources.IPPoolConverter{}, "IPPool")
+}
+
+func (syn *kubeSyncer) parseCustomK8sResourceEvent(
+	e watch.Event,
+	converter resources.CustomK8sResourceConverter,
+	resourceType string,
+) *model.KVPair {
 	// First, check the event type.
-	tpr, ok := e.Object.(*thirdparty.IpPool)
+	logContext := log.WithFields(log.Fields{
+		"ResourceType": resourceType,
+		"EventType":    e.Type,
+	})
+	tpr, ok := e.Object.(resources.CustomK8sResource)
 	if !ok {
-		log.Panicf("Invalid IPPool event. Type: %s, Object: %+v", e.Type, e.Object)
+		logContext.Panicf("Invalid custom resource event. Object: %+v", e.Object)
 	}
 
-	// Convert the received IPPool into a KVPair.
-	kvp := resources.ThirdPartyToIPPool(tpr)
+	logContext = logContext.WithField("Name", tpr.GetObjectMeta().GetName())
+	logContext.Debug("Parsing watch event")
 
-	// For deletes, we need to nil out the Value part of the KVPair
-	if e.Type == watch.Deleted {
-		kvp.Value = nil
+	// Convert the received resource into a KVPair.
+	kvp, err := converter.ToKVPair(tpr)
+	if err == nil {
+		// For deletes, we need to nil out the Value part of the KVPair
+		if e.Type == watch.Deleted {
+			kvp.Value = nil
+		}
+		return kvp
 	}
-	return kvp
+
+	// Error converting resource.  Attempt to determine the Key and treat as
```

Comment:
So the scenario this would really address is someone modifying the TPR using kubectl and creating something that did not parse correctly.

The question is what should we do?  For calico/nodes that are just starting they won't see the entry in the datastore.  For existing calico/nodes they'll still be using the previously valid instance of the resource.  A restart of a calico/node will have the effect of deleting the entry - so it feels like the syncer should therefore treat this as a delete.

WDYT?